Python's shutil.copy2 fails on Android

jerryasher opened the pull request:
Python's shutil.copy2 fails on Android when copying a file's meta data (perm bits, access times) onto certain filesystems (presumably FAT). This is documented as python [issue28141](https://bugs.python.org/issue28141)

These commits workaround that bug by

+ creating a new function copy_file_metadata in utils.py
+ wrapping calls to copy2 in a try/except clause that logs any errors that occur but keep execution going
+ changing the calls to shutil.copy2 to calls to the new function

jerryasher commented:
Thanks, I'll attend to those.

I appreciate the output of flake8, I need to run that manually. I use emacs
and have had to turn elpy-mode off, because something about elpy keeps
rewriting a lot of the whitespace of one of the two files, and I haven't
figured out yet how to turn the rewriting off.

On Wed, Sep 28, 2016 at 6:01 PM, Kevin Yap <notifications@github.com> wrote:

> *@iKevinY* requested changes on this pull request.
>
> Thanks for reopening this PR on a new branch, @jerryasher
> <https://github.com/jerryasher>! flake8 is complaining about a couple of
> things <https://travis-ci.org/getpelican/pelican/jobs/163572182#L186>;
> namely, it wants the imports in generators.py to appear in alphabetical
> order, and since your new helper function wraps shutil.copy2, shutil
> doesn't need to be imported anymore.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/getpelican/pelican/pull/2016#pullrequestreview-2057641>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAosCVF34jRgnkVHetryBz5oqItVsuNoks5quw3vgaJpZM4KJcR3>
> .
>

jerryasher commented:
@avaris I see what you're saying, the new little worker, copy_file_metadata is actually used twice within utils's copy, so maybe it makes more sense to keep that utility then duplicate the try/except.

Hey, can you explain what `def walk_error` does within the `def copy` function? Is that just a vestigial piece of code or does it have some function?

avaris commented:
Looks like `walk_error` was introduced in 0a48371. I don't know why it's there, I don't see it being used anywhere. Maybe it was part of the overhaul at some point and refactored out, but left with the stray definition. In any case, it should be safe to remove it.

jerryasher commented:
Hi @justinmayer,  I see the label "awaiting user feedback" was added to this pull request, but I guess I do not know who "the user" is. Are you waiting on me?

If so, are you waiting on me to comment on @avaris' comment?

I did move the function copy_file_metadata up to below the copy function. I did not remove walk_error, though I am happy to do so, as it is extraneous to what this change request requires. 

Tonight I

1. updated my master branch with the latest getpelican master by using `git pull`
1. updated my copy_file_metadata branch by merging my master branch into it.
1. pushing both of my branches onto my repository here.

This places two commits into this pull request. The first are my changes, the next is presumably the recent changes to getpelican/pelican master.

I am not sure if I needed to update my branch at all, or if I did it the right way. What do you think?

justinmayer commented:
Yes, I was awaiting your response to the latest comments from @Avaris. Now that you have made some follow-up changes, there are two things left to do:

First, as you can see by the indicator at the bottom of this pull request, there is a failing check. In this case, it is compliance with PyCodeStyle, formerly known as PEP 8, that needs to be addressed ([see test output](https://travis-ci.org/getpelican/pelican/jobs/165759469#L187)). You can run `flake8` locally in order to display the failing check(s), and then you can manually fix the displayed problem(s):
```
cd pelican
pip install -r requirements/developer.pip
flake8 pelican/generators.py pelican/utils.py
```
Second, once `flake8` checks pass, you should [rebase & squash your commits](https://github.com/getpelican/pelican/wiki/Git-Tips#squashing-commits) into a single commit.

jerryasher commented:
Okay, try this one!

justinmayer commented:
Appreciate for your work on this, Jerry. And thanks to @iKevinY and @avaris for reviewing.
